core/blockchain.go

```diff
@@ -895,57 +898,86 @@ func (bc *BlockChain) ValidateCanonicalChain() error {
 			}
 		}
 
-		current = bc.GetBlockByHash(current.ParentHash())
+		blkReceipts := bc.GetReceiptsByHash(current.Hash())
+		if blkReceipts.Len() != len(txs) {
```

Comment:
We should confirm that:
* `len(receipts) == len(txs)`
* for each `receipt`, there is a matching block tx with the same hash
* each `receipt` has a block hash equal to the current block and and a block index equal to the current block

These could look something like: https://gist.github.com/patrick-ogrady/34e6458e698d5db68b8567abe7e96521#file-integrity_check-go-L98-L108